### PR TITLE
Fix missing committish

### PR DESCRIPTION
**Pull request by huochunpeng**
@zkochan the current test on non-github repo didn't cover #9 because there is only one commit in the repo. If you have write access to the repo, please push one more commit up, so that 3325a3e is not master head.

```js
test('resolveFromGit() with commit from non-github repo', async (t) => {
  const resolveResult = await resolveFromGit({pref: 'git+http://ikt.pm2.io/ikt.git#3325a3e39a502418dc2e2e4bf21529cbbde96228'})
```

### CODE REVIEW

File: src/parsePref.ts [hunk 1]
```diff
@@ -113,12 +115,10 @@ function setGitCommittish (committish: string | null) {
   if (committish != null && committish.length >= 7 && committish.slice(0, 7) === 'semver:') {
     return {
       gitCommittish: null,
-      gitRange: decodeURIComponent(committish.slice(7)),
+      gitRange: committish.slice(7),
```
Comment (huochunpeng) [inline src/parsePref.ts]: decodeURIComponent is not needed for hosted git repo, as hosted-git-info did decode.

Pushed decodeURIComponent up a level, so line 121 will get the missing decode call.

File: src/parsePref.ts [hunk 2]
```diff
@@ -113,12 +115,10 @@ function setGitCommittish (committish: string | null) {
...
     }
   }
-  return {
-    gitCommittish: committish === '' ? null : committish,
-  }
+  return {gitCommittish: committish}
```
Comment (zkochan) [inline src/parsePref.ts]: if you name the variable gitCommittish, it will be nicer

```typescript
return { gitCommittish }
```

**Comment by zkochan** [comment]
I don't have access to that repo. We can use another one or create our own

File: src/parsePref.ts (hunk 2 above)
Comment (huochunpeng) [reply src/parsePref.ts]: I think it's better to keep "committish" as that's used in the two calling of this function.

**Comment by huochunpeng** [comment]
In long run, I think it's better to mock git call, and mock http head request in test. So the tests don't depend on external resources, and also we can test more edge cases.

I would also export parsePref in order to test it thoroughly. For instance, it's easier to cover #9 with unit test of parsePref.

**Comment by zkochan** [comment]
@huochunpeng, I pushed this repo to gitlab: https://gitlab.com/pnpm/git-resolver

so it can be used for testing

**Comment by huochunpeng** [comment]
Thx, but that's a gitlab repo, will go through code branch of using hosted-git-info (even when you don't use shortcut, hosted-git-info recognise it). We need another repo, not github/bitbucket/gitlab.

**Comment by zkochan** [comment]
Could we start a git server locally?

**Comment by huochunpeng** [comment]
I will try that.

BTW, can you duplicate this repo on Bitbucket too? So that tests for all hosted repos are fully controlled by you.

**Comment by zkochan** [comment]
ok, I'll do it this evening

**Comment by zkochan** [comment]
bitbucket here: https://bitbucket.org/pnpmjs/git-resolver

**Comment by huochunpeng** [comment]
A local file:// protocol did the job.

File: src/parsePref.ts [hunk 3]
```diff
@@ -25,6 +25,12 @@ const gitProtocols = new Set([
   'git+ftp',
   'git+file',
   'git+ssh',
+  'http',
+  'https',
+  'rsync',
+  'ftp',
+  'file',
```
Comment (zkochan) [inline src/parsePref.ts]: @huochunpeng
This is why tests started to fail. The file package is currently handled by `https://github.com/pnpm/local-resolver`
Is git via `file:` supported by other package managers?

Comment (huochunpeng) [reply src/parsePref.ts]: Right, I will get rid of the "file', but retain other 3. Does this sound good?

The reason I add them is that fact npm support them without git+ prefix.

Comment (zkochan) [reply src/parsePref.ts]: http and https may be problematic as well because tarball dependencies may be specified that way.

We can support it as well but it will need some refactoring and more tests in supi to handle all the different edge cases.

Comment (huochunpeng) [reply src/parsePref.ts]: Ok, I will revert all of them for now.

Comment (huochunpeng) [reply src/parsePref.ts]: I will leave "ssh" there?

Comment (huochunpeng) [reply src/parsePref.ts]: It's quite common to use ssh without prefix.

Comment (zkochan) [reply src/parsePref.ts]: ok
